src/main/java/picard/sam/markduplicates/MarkDuplicates.java

```diff
@@ -384,13 +386,12 @@ protected int doWork() {
                     }
                     final boolean isInDuplicateSet = recordInFileIndex == nextRepresentativeIndex ||
                             (sortOrder == SAMFileHeader.SortOrder.queryname &&
-                                    recordInFileIndex > nextDuplicateIndex);
+                                    recordInFileIndex > nextRepresentativeIndex && rec.getReadName().equals(representativeQueryName));
```

Comment:
can you rename this/comment the logic for "isInDuplicateSet" here to clairify this? This took a while to pick through what all of the steps are doing...

Reply:
maybe "Is RepresentativeReadForADuplicateSet"

Reply:
So this essentially says that if we are in queryname sorted mode (It should be noted this won't work if we are in query grouped mode even though the logic should be the same, maybe that check should be updated), we mark both the first read and its mates with the duplicate sets but if we are coordinate sorted the mates don't get the flag? Is that asymmetrical behavior desirable?

Reply:
Yeah this is super confusing, I actually couldn't figure out how this even works just now until I realized the variables are horribly named.

I believe that mates should also be marked with duplicate sets when the input is coordinate sorted as well (which is confirmed by our tests).  The reason is that in `addRepresentativeReadIndex` we add an entry for both reads of the fragment if `end.read1IndexInFile != end.read2IndexInFile`.  And when the ReadEnds object gets built we keep those entries the same for queryname sorted, but different for coordinate sorted (see L548 and thereabouts).  I think this really goes back to your overall point that this code has become mangled to the point obfuscation.   Which I completely agree with, but don't know that anyone has the capacity to fix, unfortunately. 

Regardless, I will try to at least make this part a little clearer